### PR TITLE
Add ST_Buffer function

### PR DESCRIPTION
**Pull request by mbasmanova**

### CODE REVIEW

File: presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 1]
```diff
@@ -113,6 +113,23 @@ public static Slice stAsText(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    {
+        if (distance < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Distance parameter for ST_Buffer must be greated then or equal to 0");
+        }
+
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.isEmpty()) {
+            return null;
+        }
+        return serialize(geometry.buffer(distance));
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: When `distance` is `0` the `buffer` method returns `this`, which is the `geometry` instance here. Do you think it is worth optimizing that case in this method to get rid of the deser/ser overhead:

```
if (distance == 0) {
    return input;
}
```

File: presto-geospatial/src/test/java/com/facebook/presto/geospatial/TestGeoQueries.java [hunk 2]
```diff
@@ -74,6 +75,29 @@ public void testSTArea()
...
+
+        // empty geometry
+        assertFunction("ST_Buffer(ST_GeometryFromText('POINT EMPTY'), 1)", GEOMETRY, null);
+
+        // negative distance
+        assertInvalidFunction("ST_Buffer(ST_Point(0, 0), -1.2)", "Distance parameter for ST_Buffer must be greated then or equal to 0");
+    }
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/test/java/com/facebook/presto/geospatial/TestGeoQueries.java]: please add tests for the cases where `distance` is: `infinity()`, `-infinity()`, and `nan()`.

File: presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 3]
```diff
@@ -113,6 +113,23 @@ public static Slice stAsText(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return Slices.utf8Slice(deserialize(input).asText());
     }
 
+    @SqlNullable
+    @Description("Returns a geometry that represents all points whose distance from specified geometry is less than or equal to specified distance")
+    @ScalarFunction("ST_Buffer")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice stBuffer(@SqlType(GEOMETRY_TYPE_NAME) Slice input, @SqlType(StandardTypes.DOUBLE) double distance)
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: we can static import `StandardTypes.DOUBLE` since we static import the other types in this class

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 4]
```diff
@@ -96,6 +96,11 @@ Operations
 
     Returns the closure of the combinatorial boundary of this geometry.
 
+.. function:: ST_Buffer(Geometry, distance) -> Geometry
+
+    Returns a geometry that represents all points whose distance from specified geometry
```
[the remark this answers is not in the document]
Comment (nezihyigitbasi) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: `the specified`

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 5]
```diff
@@ -96,6 +96,11 @@ Operations
...
+    is less than or equal to specified distance.
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: `the specified`

**Comment by mbasmanova** [comment]
@nezihyigitbasi , thank you for review. All comments addressed.

File: presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 6]
```diff
@@ -88,7 +89,7 @@ public static Slice stPolygon(@SqlType(StandardTypes.VARCHAR) Slice input)
 
     @Description("Returns the area of a polygon using Euclidean measurement on a 2D plane (based on spatial ref) in projected units")
     @ScalarFunction("ST_Area")
-    @SqlType(StandardTypes.DOUBLE)
+    @SqlType(DOUBLE)
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: It would be nice to do the static import change in another commit as that change modifies several other parts of this class.

File: presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 7]
```diff
@@ -113,6 +114,31 @@ public static Slice stAsText(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    {
+        if (distance < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "distance must be greater than or equal to 0");
+        }
+
+        if (distance == 0) {
+            return input;
+        }
+
+        if (Double.isNaN(distance)) {
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: let's move this `isNan` check above the `distance == 0` check so that all failure case checks are together and at the beginning.

File: presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 8]
```diff
@@ -113,6 +114,31 @@ public static Slice stAsText(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "distance must not be NaN");
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: `distance must not be NaN` -> `distance is NaN`

File: presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 9]
```diff
@@ -113,6 +114,31 @@ public static Slice stAsText(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
     }
 
+    @SqlNullable
+    @Description("Returns the geometry that represents all points whose distance from the specified geometry is less than or equal to the specified distance")
+    @ScalarFunction("ST_Buffer")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice stBuffer(@SqlType(GEOMETRY_TYPE_NAME) Slice input, @SqlType(DOUBLE) double distance)
+    {
+        if (distance < 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "distance must be greater than or equal to 0");
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: looking at this again let's just say `distance is negative`.

**Comment by mbasmanova** [comment]
Addressed comments. Split the change into two commits.

**Comment by nezihyigitbasi** [comment]
merged, thanks @mbasmanova!
